common/go-utils/src/inv.rs

```diff
@@ -95,28 +129,15 @@ impl GoFile {
 ///
 /// HTTP issues connecting to the upstream releases endpoint, as well
 /// as json and Go version parsing issues, will return an error.
-pub fn list_upstream_artifacts() -> Result<Vec<Artifact>, String> {
-    ureq::get(GO_RELEASES_URL)
+pub fn list_upstream_artifacts() -> Result<Vec<Artifact>, ListUpstreamArtifactsError> {
+    Ok(ureq::get(GO_RELEASES_URL)
         .call()
-        .map_err(|e| e.to_string())?
+        .map_err(|e| ListUpstreamArtifactsError::InvalidResponse(Box::new(e)))?
         .into_json::<Vec<GoRelease>>()
-        .map_err(|e| e.to_string())?
+        .map_err(ListUpstreamArtifactsError::ParseJsonResponse)?
         .iter()
         .flat_map(|release| &release.files)
-        .filter(|file| !file.sha256.is_empty())
-        .filter_map(|file| {
-            file.target_arch()
-                .filter(|target| target == ARCH)
-                .map(|target| {
-                    Version::parse_go(&file.version)
-                        .map(|version| Artifact {
-                            go_version: file.version.clone(),
-                            semantic_version: version,
-                            architecture: target,
-                            sha_checksum: file.sha256.clone(),
-                        })
-                        .map_err(|e| e.to_string())
-                })
-        })
-        .collect::<Result<Vec<_>, _>>()
+        .filter(|file| !file.sha256.is_empty() && file.os == "linux" && file.arch == "amd64")
+        .map(Artifact::try_from)
+        .collect::<Result<Vec<_>, _>>()?)
```

Comment:
since you're returning a `Result` can you not drop the `Ok(...?)` wrapper?

Reply:
It seems like this is required for `GoFileConversionError` to be mapped to `ListUpstreamArtifactsError` expected? I'm not sure if there's a better approach, but an alternative implementation of the whole method (that avoid wrapping in `Ok(...?)`, could look like this and explicitly map the error type:

```
pub fn list_upstream_artifacts() -> Result<Vec<Artifact>, ListUpstreamArtifactsError> {
    ureq::get(GO_RELEASES_URL)
        .call()
        .map_err(|e| ListUpstreamArtifactsError::InvalidResponse(Box::new(e)))?
        .into_json::<Vec<GoRelease>>()
        .map_err(ListUpstreamArtifactsError::ParseJsonResponse)?
        .iter()
        .flat_map(|release| &release.files)
        .filter(|file| !file.sha256.is_empty() && file.os == "linux" && file.arch == "amd64")
        .map(|file| Artifact::try_from(file).map_err(ListUpstreamArtifactsError::Conversion))
        .collect::<Result<Vec<_>, _>>()
}
```
Not sure this is better though - what do you think?

Reply:
I've found that (most of the time) explicit error mapping is a lot easier to follow than relying on automatic conversions.  E.g.; 
* `.map(Artifact::try_from)` is succinct but requires you to wrap the while expression in `Ok(...)?` which looks odd and it's unclear, at a glance, that it's due to this line
* `.map(|file| Artifact::try_from(file).map_err(Into::into))` is a bit better since it handles the conversion within the `map` operation which means you can drop the `Ok(...)?` but you also have to do some jumping around to figure out the `GoFileConversionError` → `ListUpstreamArtifactsError::Conversion` which gets that automatic `into` implementation since you annotated the error with `#[from]`
* `.map(|file| Artifact::try_from(file).map_err(ListUpstreamArtifactsError::Conversion))` is what I would suggest since it's very clear how the error maps at the cost of being slightly more verbose

Your choice as to which you want to use here since they're all correct. I was nudged early on by @Malax towards that last option and i've found that extra verbosity pays off when returning to previously written code.

Reply:
Ok cool -- I went with the last option in https://github.com/heroku/buildpacks-go/pull/229/commits/e5dcf1dc313ad1c02c8cffc5b994af19b63c0797 :)